lib/pbench/server/database/models/tracker.py

```diff
@@ -0,0 +1,734 @@
+                on the dataset.
+
+    The error text will identify the dataset and metadata key that was
+    specified.
+    """
+
+    def __init__(self, dataset, key):
+        super().__init__(dataset, key)
+
+    def __str__(self):
+        return f"{self.dataset} already has metadata key {self.key}"
+
+
+class States(enum.Enum):
+    """
+    States: Track the progress of a dataset (tarball) through the various
+    stages and operation of the Pbench server.
+    """
+
+    UPLOADING = ("Uploading", True)
+    UPLOADED = ("Uploaded", False)
+    UNPACKING = ("Unpacking", True)
+    UNPACKED = ("Unpacked", False)
+    INDEXING = ("Indexing", True)
+    INDEXED = ("Indexed", False)
+    EXPIRING = ("Expiring", True)
+    EXPIRED = ("Expired", False)
+    QUARANTINED = ("Quarantined", False)
+
+    def __init__(self, friendly: str, mutating: bool):
+        """
+        __init__ Extended ENUM constructor in order to add a value
+        to record whether each state is a "busy" state where some
+        component is mutating the dataset in some way, and a mixed
+        case "friendly" name for the state. "Mutating" states are
+        expected to be transient as the component should complete
+        the mutation, and should usually have "-ing" endings.
+
+        Args:
+            name (string): Friendly name for the state
+            mutating (boolean): True if a component is mutating dataset
+        """
+        self.friendly = friendly
+        self.mutating = mutating
+
+    def __str__(self):
+        """
+        Return the state's friendly name
+        """
+        return self.friendly
+
+
+class Dataset(Database.Base):
+    """
+    Identify a Pbench dataset (tarball plus metadata)
+
+    Columns:
+        id          Generated unique ID of table row
+        user        User associated with the dataset
+        controller  Name of controller node
+        name        Base name of dataset (tarball)
+        md5         The dataset MD5 hash (Elasticsearch ID)
+        created     Date the dataset was PUT to server
+        state       The current state of the dataset
+        transition  The timestamp of the last state transition
+    """
+
+    __tablename__ = "dataset"
+
+    transitions = {
+        States.UPLOADING: [States.UPLOADED, States.QUARANTINED],
+        States.UPLOADED: [States.UNPACKING, States.QUARANTINED],
+        States.UNPACKING: [States.UNPACKED, States.QUARANTINED],
+        States.UNPACKED: [States.INDEXING, States.QUARANTINED],
+        States.INDEXING: [States.INDEXED, States.QUARANTINED],
+        States.INDEXED: [States.INDEXING, States.EXPIRING, States.QUARANTINED],
+        States.EXPIRING: [States.EXPIRED, States.INDEXED, States.QUARANTINED],
+        # NOTE: EXPIRED and QUARANTINED do not appear as keys in the table
+        # because they're terminal states that cannot be exited.
+    }
+
+    ANTI_USER = "public"  # The "user" that owns unowned datasets
```

Comment:
In this context, it doesn't need to be present because nobody's looking it up. Exactly like the "admin" user we talk about. In practice, we'll probably want to represent "admin" at least in the user DB. I'm less sure about "public", and that's one of the reasons I renamed from `DEFAULT_USERNAME` to `ANTI_USER` ... it really means there __is__ no user. You can never authenticate as "public" because there's no such user. If we stored it in the user database we'd need some metadata to record that it can never be logged-in or edited, and I'm not sure there'd be any point.

Except, perhaps, to ensure that nobody could create a user named `public` and cause confusion... so maybe we will want a non-user user as a placeholder. I talked a bit about an install-time utility we'd run to create the Postgres tables and probably populate an "admin" user... and maybe a "public" user as well. In theory, we could even allow dynamic configuration of the "anti-user" name, though I'm not sure it matters.

Reply:
okay sure, I thought public would be an actual user we would need to create and maintain which isn't desirable. So an ANTI_USER is basically no user and you are representing that notion as a "public" user, I would say we can have ANTI_USER as None instead of naming explicitly as public and we could achieve the same result but that's nit

Reply:
That's an excellent point, Nikhil.

The "admin" user and the "public" user are different from each other, as we expect someone to log into the former, but we do not want anyone to log into the latter.  We presumably want to protect each of them (at least until "admin" graduates from a "user" to a "role"), and, if we have an actual value for the "pubic" user username, then we need to ensure that no one can create that username and thereby "claim" that data.  Tagging that data in a way which associates it with a null username seems very promising, as that's a value which we presumably won't allow to be used for a user.  (And, as you say, it saves us from all the management cruft.)

However, perhaps it would be best to capture this in an issue and handle it in another PR.

Reply:
Hmm. Yes, perhaps setting the Elasticsearch authorization to `{"user": "", "access": "public"}` would work. I admit that I hadn't considered that possibility. That'll mess with gold test output, and doing it in a separate PR makes more sense; but I'm willing to try it out.